util/merkle/src/triedbmut.rs

```diff
@@ -0,0 +1,539 @@
+        let t = TrieDBMut::new(&mut memdb, &mut root);
+        assert_eq!(t.get(&[0x5]), Ok(None));
+    }
+
+    #[test]
+    fn test_trie_existing() {
+        let mut root = H256::new();
+        let mut db = MemoryDB::new();
+        {
+            let mut t = TrieDBMut::new(&mut db, &mut root);
+            t.insert(&[0x01u8, 0x23], &[0x01u8, 0x23]).unwrap();
+        }
+
+        {
+            let t = TrieDBMut::from_existing(&mut db, &mut root); // Why can't I use '?' behind this evaluation
+            assert_eq!(t.unwrap().get(&[0x01u8, 0x23]).unwrap().unwrap(), DBValue::from_slice(&[0x01u8, 0x23]));
+        }
+    }
+
+    #[test]
+    #[ignore]
+    fn insert_empty() {
+        let mut seed = H256::new();
+        let x = StandardMap {
+            alphabet: Alphabet::Custom(b"@QWERTYUIOPASDFGHJKLZXCVBNM[/]^_".to_vec()),
+            min_key: 5,
+            journal_key: 0,
+            value_mode: ValueMode::Index,
+            count: 4,
+        }.make_with(&mut seed);
+
+        let mut db = MemoryDB::new();
+        let mut root = H256::new();
+        let mut t = TrieDBMut::new(&mut db, &mut root);
+        for &(ref key, ref value) in &x {
+            t.insert(key, value).unwrap();
+        }
+
+
+        for &(ref key, _) in &x {
+            t.insert(key, &[]).unwrap();
+        }
+
+        assert!(t.is_empty());
+        assert_eq!(*t.root(), BLAKE_NULL_RLP);
+    }
+
+    #[test]
+    fn random_insert_remove() {
+        let mut seed = H256::new();
+        let x = StandardMap {
+            alphabet: Alphabet::Custom(b"@QWERTYUIOPASDFGHJKLZXCVBNM[/]^_".to_vec()),
+            min_key: 5,
+            journal_key: 0,
+            value_mode: ValueMode::Index,
+            count: 1000,
+        }.make_with(&mut seed);
+
+        let mut db = MemoryDB::new();
+        let mut root = H256::new();
+        let mut t = TrieDBMut::new(&mut db, &mut root);
+        for &(ref key, ref value) in &x {
+            t.insert(key, value).unwrap();
+        }
+
+        for &(ref key, _) in &x {
+            t.remove(key).unwrap();
+        }
+
+        assert!(t.is_empty());
+        assert_eq!(*t.root(), BLAKE_NULL_RLP);
+    }
+
+    #[test]
+    fn return_old_values() {
+        let mut seed = H256::new();
+        let x = StandardMap {
+            alphabet: Alphabet::Custom(b"@QWERTYUIOPASDFGHJKLZXCVBNM[/]^_".to_vec()),
+            min_key: 5,
+            journal_key: 0,
+            value_mode: ValueMode::Index,
+            count: 4,
+        }.make_with(&mut seed);
+
+        let mut db = MemoryDB::new();
+        let mut root = H256::new();
+        let mut t = TrieDBMut::new(&mut db, &mut root);
+        for &(ref key, ref value) in &x {
+            assert!(t.insert(key, value).unwrap().is_none());
+            assert_eq!(t.insert(key, value).unwrap(), Some(DBValue::from_slice(value)));
+        }
+
+        for (key, value) in x {
+            assert_eq!(t.remove(&key).unwrap(), Some(DBValue::from_slice(&value)));
```

Comment:
```rust
assert_eq!(Ok(Some(DBValue::from_slice(&value))), t.remove(&key));
```

Reply:
Two code works same when the test passes, but it gives more information when the test fails.

Reply:
when I test with original one, i got this error
```
failures:

---- triedbmut::tests::return_old_values stdout ----
thread 'triedbmut::tests::return_old_values' panicked at 'assertion failed: `(left == right)`
  left: `None`,
 right: `Some([128])`', util/merkle/src/triedbmut.rs:537:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```
and using code you gave me, I got similar one
```
failures:

---- triedbmut::tests::return_old_values stdout ----
thread 'triedbmut::tests::return_old_values' panicked at 'assertion failed: `(left == right)`
  left: `Ok(Some([128]))`,
 right: `Ok(None)`', util/merkle/src/triedbmut.rs:537:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```
Could you explain why you recommend ```assert_eq!(Ok(Some(DBValue::from_slice(&value))), t.remove(&key));``` or you can give me a keywords to search

Reply:
When `remove` returns `Err(_)`, the assetion is in `Result` so it doesn't print `util/merkle/src/triedbmut.rs:537:13` but prints `some path to result.rs:...:..`.